examples/ota-provider-app/linux/main.cpp

```diff
@@ -0,0 +1,149 @@
+
+#include "BdxOtaSender.h"
+#include "OTAProviderExample.h"
+
+#include <fstream>
+#include <iostream>
+
+using namespace chip;
+using namespace chip::ArgParser;
+
+using chip::app::clusters::OTAProviderDelegate;
+
+constexpr uint16_t kOptionFilepath = 'f';
+
+const char * gOtaFilepath = nullptr;
+
+// Arbitrary BDX Transfer Params
+constexpr uint32_t kMaxBdxBlockSize = 1024;
+constexpr uint32_t kBdxTimeoutMs    = 5 * 60 * 1000; // OTA Spec mandates >= 5 minutes
+constexpr uint32_t kBdxPollFreqMs   = 500;
+
+bool HandleOptions(const char * aProgram, OptionSet * aOptions, int aIdentifier, const char * aName, const char * aValue)
+{
+    bool retval = true;
+    std::ifstream ifs;
+
+    switch (aIdentifier)
+    {
+
+    case kOptionFilepath:
+        ifs.open(aValue, std::ifstream::in);
+        if (!ifs.good())
+        {
+            PrintArgError("%s: failed to open %s\n", aProgram, aValue);
+            retval = false;
+        }
+        else
+        {
+            gOtaFilepath = aValue;
+        }
+        break;
+
+    default:
+        PrintArgError("%s: INTERNAL ERROR: Unhandled option: %s\n", aProgram, aName);
+        retval = false;
+        break;
+    }
+
+    return (retval);
+}
+
+OptionDef cmdLineOptionsDef[] = {
+    { "filepath", kArgumentRequired, kOptionFilepath },
+    {},
+};
+
+OptionSet cmdLineOptions = { HandleOptions, cmdLineOptionsDef, "PROGRAM OPTIONS",
+                             "  -f <file>\n"
+                             "  --filepath <file>\n"
+                             "        Path to a file containing an OTA image.\n" };
+
+HelpOptions helpOptions("ota-provider-app", "Usage: ota-provider-app [options]", "1.0");
+
+OptionSet * allOptions[] = { &cmdLineOptions, &helpOptions, nullptr };
+
+int main(int argc, char * argv[])
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    OTAProviderExample otaProvider;
+    BdxOtaSender bdxServer;
+    Messaging::ExchangeManager * exchangeMgr;
+
+    if (chip::Platform::MemoryInit() != CHIP_NO_ERROR)
+    {
+        fprintf(stderr, "FAILED to initialize memory\n");
+        return 1;
+    }
+
+    if (DeviceLayer::PlatformMgr().InitChipStack() != CHIP_NO_ERROR)
+    {
+        fprintf(stderr, "FAILED to initialize chip stack\n");
+        return 1;
+    }
+
+    if (!chip::ArgParser::ParseArgs(argv[0], argc, argv, allOptions))
+    {
+        return 1;
+    }
+
+    DeviceLayer::ConfigurationMgr().LogDeviceConfig();
+    InitServer();
+
+    exchangeMgr = chip::ExchangeManager();
+    err         = exchangeMgr->RegisterUnsolicitedMessageHandlerForProtocol(Protocols::BDX::Id, &bdxServer);
+
+    ChipLogDetail(SoftwareUpdate, "using OTA file: %s", gOtaFilepath ? gOtaFilepath : "(none)");
+
+    if (gOtaFilepath != nullptr)
+    {
+        otaProvider.SetOTAFilePath(gOtaFilepath, strlen(gOtaFilepath));
+        bdxServer.SetFilepath(gOtaFilepath, strlen(gOtaFilepath));
+    }
+
+    // TODO: is there any way to not hardcode the endpoint id?
```

Comment:
Open question: is there an API that can return the endpoint ID that OTA Provider is on?